rust/error-tracking/src/team_cache.rs

```diff
@@ -0,0 +1,58 @@
+use std::{future::Future, time::Duration};
+
+use common_types::Team;
+use moka::sync::{Cache, CacheBuilder};
+use tokio::sync::RwLock;
+
+use crate::error::Error;
+
+// This /could/ be moved into common/types, but I'd have to take
+// a dependency on tokio and moka to do it, and I don't want to
+// do that just yet (although it is used across here and feature
+// flags, so....)
+pub struct TeamCache {
+    // The lock here isn't necessary (the Cache is concurrent),
+    // but is used to ensure the DB is only hit once.
+    // Note that we cache the none-case to prevent
+    // people hammering us with false tokens and bringing
+    // down PG.
+    teams: RwLock<Cache<String, Option<Team>>>,
+}
+
+impl TeamCache {
+    pub fn new(capacity: u64, ttl_seconds: u64) -> Self {
+        let cache = CacheBuilder::new(capacity)
+            .time_to_live(Duration::from_secs(ttl_seconds))
+            .build();
+
+        Self {
+            teams: RwLock::new(cache),
+        }
+    }
+
+    pub async fn get_or_insert_with<F>(&self, token: &str, f: F) -> Result<Option<Team>, Error>
+    where
+        F: Future<Output = Result<Option<Team>, Error>>,
+    {
+        let teams = self.teams.read().await;
+        if let Some(team) = teams.get(token) {
+            return Ok(team.clone());
+        }
```

Comment:
Why do we do the read lock first if we try to get the value again with the write lock on line 43

Reply:
It's a happy path optimisation - if we have a cached value, we only ever take the read lock, which is lighter weight because multiple readers can concurrently hold it to get a copy of the value. If we end up needing to modify the cache (to insert a new value), we have to claim the lock exclusively (can't have concurrent readers and writers), which takes longer/blocks all other accessing threads.

This could be a `Mutex`, and then we're just take a full exclusive lock every time, but with caches you expect writes to be rare and reads to be common, so this "try the happy path then bail off to the slow one if you have to" pattern is common.

Reply:
Makes total sense. Is the second `get` after taking the write lock to protect against an insert between the original read given the await could delay the function execution?

Reply:
Yep, exactly - although the await isn't the only reason you could have an update between taking the read lock and taking the write lock - the async runtime we use, tokio, is multithreaded, so two futures can be executing in parallel, and could race on updating the cache even if we didn't have an await point here